proposals/NNNN-inheritance-of-actor-isolation.md

```diff
@@ -0,0 +1,551 @@
+callee is a little trickier.  If isolation is specified as a parameter,
+then the caller must implicitly provide an argument to it; the most
+obvious way to do that is to create a new special form for default
+arguments, like `#line`, which expands to an expression that
+evaluates to the isolation of the caller.
+
+### Generalized `isolated` parameters
+
+The type of an `isolated` parameter must be an *isolation type*.
+Currently, the only kind of isolation is a possibly-optional actor type,
+which is to say, either `T` or `Optional<T>`, where `T` either conforms
+to `Actor` or is a protocol type that implies `Actor`.
+
+If a function's `isolated` parameter has an optional actor type, then
+the dynamic isolation of the function depends on whether the argument
+value is `nil`.  If it is `nil`, then the function behaves dynamically
+as it were non-isolated; for example, if the function is `async`, it
+resets isolation on entry under [SE-0338][] just as a non-isolated
+function would.  Otherwise, the function behaves dynamically as it
+were isolated to the unwrapped actor reference.
+
+According to [SE-0304][SE-0304-propagation], closures passed directly
+to the `Task` initializer (i.e. `Task { /*here*/ }`) inherit the
+statically-specified isolation of the current context if:
+
+- the current context is non-isolated,
+- the current context is isolated to a global actor, or
+- the current context has an `isolated` parameter (including the
+  implicit `self` of an actor method) and that parameter is strongly
+  captured by the closure.
+
+The third clause is modified by this proposal to say that isolation
+is also inherited if a non-optional binding of an isolated parameter
+is captured by the closure (see below).
+
+### Generalized isolation checking
+
+When calling a function with an `isolated` parameter, the function
+shares the same isolation as the current context if:
+
+- the current context is non-isolated, the parameter type is optional,
+  and the argument expression is `nil` or a reference to `Optional.none`;
+
+- the current context has an `isolated` parameter (including the
+  implicitly-`isolated` `self` parameter of an actor function) and
+  the argument expression is a reference to that parameter or a
+  non-optional derivation of it (see below); or
+
+- the current context is isolated to a global actor type `T` and the
+  argument expression is `T.shared`, where `shared` is `GlobalActor`'s
+  protocol requirement or the concrete declaration which provides it
+  in `T`'s conformance to `GlobalActor`.
+
+An expression is a non-optional derivation of an isolated parameter
+`param` if it is:
+- `param?` (the optional-chaining operator);
+- `param!` (the force-unwrapping operator); or
+- a reference to a *non-optional binding* of `param`, i.e. a `let`
+  constant initialized by a successful pattern-match which removes
+  the optionality from `param`, such as `ref` in `if let ref = param`.
```

Comment:
No, because the derivation must be non-optional. You might imagine supporting something like this:

```swift
func test(param: isolated (any Actor)?) {
  if let value = param ?? nil {
    // 'value' could be considered a non-optional derivation of 'param' here
  }
}
```

But the proposal does not support looking through any operators other than `?` and `!`.

Reply:
Sorry, poorly phrased; consider the question to apply to line 249—

Given the following:

```swift
func f(_ param: isolated (any Actor)?) {
  g(<<< expression >>>)
}
func g(_ param: isolated (any Actor)?) { }
```

...the proposal states that `g` shares an isolation context if `<<< expression >>>` is `param` or `param!`, the latter because it is a "non-optional derivation" which is then implicitly promoted back to an `Optional` type. For completeness (cf. note 5 about pointlessness), what about `param ?? nil`.

Reply:
Ah, I see your point, thank you for elaborating! Still the answer is that only the forms mentioned in the proposal are supported, so that does not include `param ?? nil`. I'm reluctant to start adding other operators to the list, because this will not generalize to arbitrary operators and functions. I think whether or not that's the right choice is fair to raise during the review.